File: src/main/java/com/tkr/desk/CounterpartyLimit.java
package com.tkr.desk;

import com.tkr.types.WireTypes.*;
import com.tkr.types.WireTypes.Status;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Counterparty exposure limits with notional, quantity, turnover checks.
 * Ported from legacy-cpp counterparty_limit.cc.
 */
public final class CounterpartyLimit {

    public enum LimitKind { NOTIONAL, QUANTITY, DAILY_TURNOVER, NET_EXPOSURE, OPEN_ORDERS }

    public enum LimitScope { PER_COUNTERPARTY, PER_SYMBOL, PER_ACCOUNT, DESK_WIDE }

    public static final class CounterpartyProfile {
        public int counterpartyId;
        public String name = "";
        public int creditRatingBp;
        public boolean approved = true;
        public int parentId;
    }

    public static final class ExposureLimit {
        public LimitKind kind;
        public LimitScope scope;
        public int limitId;
        public int counterpartyId;
        public int symbolId;
        public int accountId;
        public long maxValue;
        public long warningThreshold;
        public boolean hardBlock = true;
    }

    public static final class ExposureReading {
        public int counterpartyId;
        public int accountId;
        public int symbolId;
        public long notionalCents;
        public long qtyMilli;
        public int openOrderCount;
        public long dailyTurnoverCents;
        public long netExposureCents;
    }

    public static final class LimitBreach {
        public LimitKind kind;
        public int limitId;
        public int counterpartyId;
        public int recordId;
        public long currentValue;
        public long limitValue;
        public boolean isWarning;
        public String reason = "";
    }

    public static final class CounterpartyLimitConfig {
        public int deskId;
        public boolean enforceHardBlocks;
        public boolean aggregateParent;

        public CounterpartyLimitConfig(int deskId, boolean enforceHardBlocks) {
            this.deskId = deskId;
            this.enforceHardBlocks = enforceHardBlocks;
            this.aggregateParent = false;
        }
    }

    public static final class CounterpartyLimitResult {
        public Status status = Status.OK;
        public List<LimitBreach> breaches = new ArrayList<>();
        public List<ExposureReading> readings = new ArrayList<>();
        public int breachCount;
        public int warningCount;
    }

    private static final int COUNTERPARTY_BASE = 50000;

    private final CounterpartyLimitConfig config;
    private final Map<Integer, CounterpartyProfile> counterparties = new HashMap<>();
    private final List<ExposureLimit> limits = new ArrayList<>();
    private final Map<Integer, ExposureReading> runningExposure = new HashMap<>();

    public CounterpartyLimit() {
        this(new CounterpartyLimitConfig(0, true));
    }

    public CounterpartyLimit(CounterpartyLimitConfig config) {
        this.config = config != null ? config : new CounterpartyLimitConfig(0, true);
        loadDefaultLimits();
    }

    public void loadDefaultLimits() {
        CounterpartyProfile brokerA = new CounterpartyProfile();
        brokerA.counterpartyId = 50001;
        brokerA.name = "BrokerAlpha";
        brokerA.creditRatingBp = 8500;
        brokerA.approved = true;
        registerCounterparty(brokerA);

        CounterpartyProfile brokerB = new CounterpartyProfile();
        brokerB.counterpartyId = 50002;
        brokerB.name = "BrokerBeta";
        brokerB.creditRatingBp = 7200;
        brokerB.approved = true;
        registerCounterparty(brokerB);

        ExposureLimit notional = new ExposureLimit();
        notional.kind = LimitKind.NOTIONAL;
        notional.scope = LimitScope.PER_COUNTERPARTY;
        notional.limitId = 6001;
        notional.maxValue = 500_000_000L;
        notional.warningThreshold = 400_000_000L;
        notional.hardBlock = true;
        setLimit(notional);

        ExposureLimit qty = new ExposureLimit();
        qty.kind = LimitKind.QUANTITY;
        qty.scope = LimitScope.PER_SYMBOL;
        qty.limitId = 6002;
        qty.maxValue = 10_000_000L;
        qty.warningThreshold = 8_000_000L;
        qty.hardBlock = true;
        setLimit(qty);

        ExposureLimit turnover = new ExposureLimit();
        turnover.kind = LimitKind.DAILY_TURNOVER;
        turnover.scope = LimitScope.PER_COUNTERPARTY;
        turnover.limitId = 6003;
        turnover.maxValue = 1_000_000_000L;
        turnover.warningThreshold = 800_000_000L;
        turnover.hardBlock = false;
        setLimit(turnover);
    }

    public void registerCounterparty(CounterpartyProfile profile) {
        counterparties.put(profile.counterpartyId, profile);
    }

    public void setLimit(ExposureLimit limit) {
        for (int i = 0; i < limits.size(); i++) {
            if (limits.get(i).limitId == limit.limitId) {
                limits.set(i, limit);
                return;
            }
        }
        limits.add(limit);
    }

    public void removeLimit(int limitId) {
        limits.removeIf(l -> l.limitId == limitId);
    }

    public void clearLimits() {
        limits.clear();
        runningExposure.clear();
    }

    public static long computeRecordNotional(WireBatchRecord rec) {
        return (long) rec.qtyMilli * rec.priceTick / 1000L;
    }

    public static int deriveCounterpartyId(int accountId) {
        return COUNTERPARTY_BASE + (accountId % 100);
    }

    public boolean isCounterpartyApproved(int counterpartyId) {
        CounterpartyProfile profile = counterparties.get(counterpartyId);
        return profile == null || profile.approved;
    }

    public long lookupCurrentExposure(int counterpartyId, LimitKind kind) {
        ExposureReading reading = runningExposure.get(counterpartyId);
        if (reading == null) return 0;
        return switch (kind) {
            case NOTIONAL -> reading.notionalCents;
            case QUANTITY -> reading.qtyMilli;
            case DAILY_TURNOVER -> reading.dailyTurnoverCents;
            case NET_EXPOSURE -> reading.netExposureCents;
            default -> 0;
        };
    }

    public ExposureReading buildReading(WireBatchRecord rec) {
        ExposureReading reading = new ExposureReading();
        reading.counterpartyId = deriveCounterpartyId(rec.accountId);
        reading.accountId = rec.accountId;
        reading.symbolId = rec.symbolId;
        reading.notionalCents = computeRecordNotional(rec);
        reading.qtyMilli = rec.qtyMilli;
        reading.openOrderCount = 1;
        reading.dailyTurnoverCents = reading.notionalCents;
        reading.netExposureCents = reading.notionalCents;
        return reading;
    }

    public ExposureReading aggregateReadings(List<ExposureReading> readings, int counterpartyId) {
        ExposureReading agg = new ExposureReading();
        agg.counterpartyId = counterpartyId;
        for (ExposureReading r : readings) {
            if (r.counterpartyId != counterpartyId) continue;
            agg.notionalCents += r.notionalCents;
            agg.qtyMilli += r.qtyMilli;
            agg.openOrderCount += r.openOrderCount;
            agg.dailyTurnoverCents += r.dailyTurnoverCents;
            agg.netExposureCents += r.netExposureCents;
        }
        return agg;
    }

    public List<ExposureLimit> limitsForCounterparty(int counterpartyId) {
        List<ExposureLimit> result = new ArrayList<>();
        for (ExposureLimit limit : limits) {
            if (limit.counterpartyId == 0 || limit.counterpartyId == counterpartyId) result.add(limit);
        }
        return result;
    }

    public Status checkLimit(ExposureLimit limit, ExposureReading reading, int recordId, LimitBreach out) {
        if (out == null) return Status.BOUNDS_ERROR;
        long current = switch (limit.kind) {
            case NOTIONAL -> reading.notionalCents;
            case QUANTITY -> reading.qtyMilli;
            case DAILY_TURNOVER -> reading.dailyTurnoverCents;
            case NET_EXPOSURE -> reading.netExposureCents;
            case OPEN_ORDERS -> reading.openOrderCount;
        };
        if (limit.scope == LimitScope.PER_SYMBOL && limit.symbolId != 0 && limit.symbolId != reading.symbolId) {
            return Status.OK;
        }
        if (limit.scope == LimitScope.PER_ACCOUNT && limit.accountId != 0 && limit.accountId != reading.accountId) {
            return Status.OK;
        }
        long prior = lookupCurrentExposure(reading.counterpartyId, limit.kind);
        current += prior;
        if (current > limit.maxValue) {
            out.kind = limit.kind;
            out.limitId = limit.limitId;
            out.counterpartyId = reading.counterpartyId;
            out.recordId = recordId;
            out.currentValue = current;
            out.limitValue = limit.maxValue;
            out.isWarning = false;
            out.reason = "limit breach: current " + current + " exceeds max " + limit.maxValue;
            return limit.hardBlock ? Status.COMPLIANCE_REJECT : Status.OK;
        }
        if (current > limit.warningThreshold) {
            out.kind = limit.kind;
            out.limitId = limit.limitId;
            out.counterpartyId = reading.counterpartyId;
            out.recordId = recordId;
            out.currentValue = current;
            out.limitValue = limit.warningThreshold;
            out.isWarning = true;
            out.reason = "limit warning: approaching threshold";
        }
        return Status.OK;
    }

    public CounterpartyLimitResult evaluateReading(ExposureReading reading) {
        CounterpartyLimitResult result = new CounterpartyLimitResult();
        result.status = Status.OK;
        if (!isCounterpartyApproved(reading.counterpartyId)) {
            LimitBreach breach = new LimitBreach();
            breach.kind = LimitKind.NOTIONAL;
            breach.counterpartyId = reading.counterpartyId;
            breach.reason = "counterparty not approved";
            result.breaches.add(breach);
            result.breachCount++;
            result.status = Status.COMPLIANCE_REJECT;
            return result;
        }
        for (ExposureLimit limit : limitsForCounterparty(reading.counterpartyId)) {
            LimitBreach breach = new LimitBreach();
            Status st = checkLimit(limit, reading, 0, breach);
            if (st == Status.COMPLIANCE_REJECT) {
                result.breaches.add(breach);
                result.breachCount++;
                if (config.enforceHardBlocks) result.status = Status.COMPLIANCE_REJECT;
            } else if (breach.isWarning) {
                result.breaches.add(breach);
                result.warningCount++;
            }
        }
        ExposureReading running = runningExposure.computeIfAbsent(reading.counterpartyId, id -> new ExposureReading());
        running.counterpartyId = reading.counterpartyId;
        running.notionalCents += reading.notionalCents;
        running.qtyMilli += reading.qtyMilli;
        running.dailyTurnoverCents += reading.dailyTurnoverCents;
        running.netExposureCents += reading.netExposureCents;
        running.openOrderCount += reading.openOrderCount;
        result.readings.add(reading);
        return result;
    }

    public CounterpartyLimitResult evaluate(BatchWireFrame frame) {
        CounterpartyLimitResult result = new CounterpartyLimitResult();
        result.status = Status.OK;
        if (frame == null) return result;
        for (WireBatchRecord rec : frame.records) {
            ExposureReading reading = buildReading(rec);
            CounterpartyLimitResult partial = evaluateReading(reading);
            for (LimitBreach b : partial.breaches) {
                result.breaches.add(b);
                if (b.isWarning) result.warningCount++;
                else result.breachCount++;
            }
            if (partial.status == Status.COMPLIANCE_REJECT && config.enforceHardBlocks) {
                result.status = Status.COMPLIANCE_REJECT;
            }
            result.readings.add(reading);
        }
        if (result.breachCount > 0 && config.enforceHardBlocks) result.status = Status.COMPLIANCE_REJECT;
        return result;
    }

    public void resetRunningExposure() {
        runningExposure.clear();
    }

    public long totalDeskExposure(LimitKind kind) {
        long total = 0;
        for (ExposureReading reading : runningExposure.values()) {
            total += switch (kind) {
                case NOTIONAL -> reading.notionalCents;
                case QUANTITY -> reading.qtyMilli;
                case DAILY_TURNOVER -> reading.dailyTurnoverCents;
                default -> 0;
            };
        }
        return total;
    }

    public CounterpartyLimitResult evaluateParentAggregate(BatchWireFrame frame) {
        CounterpartyLimitResult result = new CounterpartyLimitResult();
        result.status = Status.OK;
        if (!config.aggregateParent) return evaluate(frame);
        Map<Integer, List<ExposureReading>> byParent = new HashMap<>();
        for (WireBatchRecord rec : frame.records) {
            ExposureReading reading = buildReading(rec);
            int parentId = reading.counterpartyId;
            CounterpartyProfile profile = counterparties.get(reading.counterpartyId);
            if (profile != null && profile.parentId != 0) parentId = profile.parentId;
            byParent.computeIfAbsent(parentId, k -> new ArrayList<>()).add(reading);
        }
        for (Map.Entry<Integer, List<ExposureReading>> entry : byParent.entrySet()) {
            ExposureReading agg = aggregateReadings(entry.getValue(), entry.getKey());
            CounterpartyLimitResult partial = evaluateReading(agg);
            for (LimitBreach b : partial.breaches) {
                result.breaches.add(b);
                if (b.isWarning) result.warningCount++;
                else result.breachCount++;
            }
            result.readings.add(agg);
        }
        if (result.breachCount > 0 && config.enforceHardBlocks) result.status = Status.COMPLIANCE_REJECT;
        return result;
    }

    public List<LimitBreach> breachesForCounterparty(CounterpartyLimitResult result, int counterpartyId) {
        List<LimitBreach> filtered = new ArrayList<>();
        for (LimitBreach b : result.breaches) {
            if (b.counterpartyId == counterpartyId) filtered.add(b);
        }
        return filtered;
    }

    public long exposureUtilizationBp(int counterpartyId, LimitKind kind) {
        List<ExposureLimit> applicable = limitsForCounterparty(counterpartyId);
        long current = lookupCurrentExposure(counterpartyId, kind);
        for (ExposureLimit limit : applicable) {
            if (limit.kind == kind && limit.maxValue > 0) {
                return current * 10000L / limit.maxValue;
            }
        }
        return 0;
    }

    public boolean isWithinWarningThreshold(int counterpartyId, LimitKind kind) {
        long util = exposureUtilizationBp(counterpartyId, kind);
        return util > 8000 && util <= 10000;
    }

    public String formatBreach(LimitBreach breach) {
        return breach.kind + " cp=" + breach.counterpartyId + " current=" + breach.currentValue
                + " limit=" + breach.limitValue + " warn=" + breach.isWarning;
    }
}

File: legacy-cpp/src/desk/counterparty_limit.cc
#include "tkr/desk/counterparty_limit.h"

#include <algorithm>

namespace tkr {
namespace desk {
namespace {

constexpr std::uint32_t kCounterpartyBase = 50000;

}  // namespace

CounterpartyLimit::CounterpartyLimit(CounterpartyLimitConfig config)
    : config_(config) {
  LoadDefaultLimits();
}

void CounterpartyLimit::LoadDefaultLimits() {
  CounterpartyProfile broker_a{};
  broker_a.counterparty_id = 50001;
  broker_a.name = "BrokerAlpha";
  broker_a.credit_rating_bp = 8500;
  broker_a.approved = true;
  broker_a.parent_id = 0;
  RegisterCounterparty(broker_a);

  CounterpartyProfile broker_b{};
  broker_b.counterparty_id = 50002;
  broker_b.name = "BrokerBeta";
  broker_b.credit_rating_bp = 7200;
  broker_b.approved = true;
  broker_b.parent_id = 0;
  RegisterCounterparty(broker_b);

  ExposureLimit notional{};
  notional.kind = LimitKind::kNotional;
  notional.scope = LimitScope::kPerCounterparty;
  notional.limit_id = 6001;
  notional.counterparty_id = 0;
  notional.max_value = 500000000;
  notional.warning_threshold = 400000000;
  notional.hard_block = true;
  SetLimit(notional);

  ExposureLimit qty{};
  qty.kind = LimitKind::kQuantity;
  qty.scope = LimitScope::kPerSymbol;
  qty.limit_id = 6002;
  qty.counterparty_id = 0;
  qty.max_value = 10000000;
  qty.warning_threshold = 8000000;
  qty.hard_block = true;
  SetLimit(qty);

  ExposureLimit turnover{};
  turnover.kind = LimitKind::kDailyTurnover;
  turnover.scope = LimitScope::kPerCounterparty;
  turnover.limit_id = 6003;
  turnover.counterparty_id = 0;
  turnover.max_value = 1000000000;
  turnover.warning_threshold = 800000000;
  turnover.hard_block = false;
  SetLimit(turnover);
}

void CounterpartyLimit::RegisterCounterparty(const CounterpartyProfile& profile) {
  counterparties_[profile.counterparty_id] = profile;
}

void CounterpartyLimit::SetLimit(const ExposureLimit& limit) {
  for (ExposureLimit& existing : limits_) {
    if (existing.limit_id == limit.limit_id) {
      existing = limit;
      return;
    }
  }
  limits_.push_back(limit);
}

void CounterpartyLimit::RemoveLimit(std::uint32_t limit_id) {
  limits_.erase(std::remove_if(limits_.begin(), limits_.end(),
                               [limit_id](const ExposureLimit& l) {
                                 return l.limit_id == limit_id;
                               }),
                limits_.end());
}

void CounterpartyLimit::ClearLimits() {
  limits_.clear();
  running_exposure_.clear();
}

std::int64_t CounterpartyLimit::ComputeRecordNotional(
    const WireBatchRecord& rec) {
  return (static_cast<std::int64_t>(rec.qty_milli) *
          static_cast<std::int64_t>(rec.price_tick)) /
         1000;
}

std::uint32_t CounterpartyLimit::DeriveCounterpartyId(
    std::uint32_t account_id) {
  return kCounterpartyBase + (account_id % 100);
}

bool CounterpartyLimit::IsCounterpartyApproved(
    std::uint32_t counterparty_id) const {
  const auto it = counterparties_.find(counterparty_id);
  if (it == counterparties_.end()) {
    return true;
  }
  return it->second.approved;
}

std::int64_t CounterpartyLimit::LookupCurrentExposure(
    std::uint32_t counterparty_id, LimitKind kind) const {
  const auto it = running_exposure_.find(counterparty_id);
  if (it == running_exposure_.end()) {
    return 0;
  }
  switch (kind) {
    case LimitKind::kNotional:
      return it->second.notional_cents;
    case LimitKind::kQuantity:
      return it->second.qty_milli;
    case LimitKind::kDailyTurnover:
      return it->second.daily_turnover_cents;
    case LimitKind::kNetExposure:
      return it->second.net_exposure_cents;
    default:
      return 0;
  }
}

ExposureReading CounterpartyLimit::BuildReading(
    const WireBatchRecord& rec) const {
  ExposureReading reading{};
  reading.counterparty_id = DeriveCounterpartyId(rec.account_id);
  reading.account_id = rec.account_id;
  reading.symbol_id = rec.symbol_id;
  reading.notional_cents = ComputeRecordNotional(rec);
  reading.qty_milli = static_cast<std::int64_t>(rec.qty_milli);
  reading.open_order_count = 1;
  reading.daily_turnover_cents = reading.notional_cents;
  reading.net_exposure_cents = reading.notional_cents;
  return reading;
}

ExposureReading CounterpartyLimit::AggregateReadings(
    const std::vector<ExposureReading>& readings,
    std::uint32_t counterparty_id) const {
  ExposureReading agg{};
  agg.counterparty_id = counterparty_id;

  for (const ExposureReading& r : readings) {
    if (r.counterparty_id != counterparty_id) {
      continue;
    }
    agg.notional_cents += r.notional_cents;
    agg.qty_milli += r.qty_milli;
    agg.open_order_count += r.open_order_count;
    agg.daily_turnover_cents += r.daily_turnover_cents;
    agg.net_exposure_cents += r.net_exposure_cents;
  }
  return agg;
}

std::vector<ExposureLimit> CounterpartyLimit::LimitsForCounterparty(
    std::uint32_t counterparty_id) const {
  std::vector<ExposureLimit> result;
  for (const ExposureLimit& limit : limits_) {
    if (limit.counterparty_id == 0 ||
        limit.counterparty_id == counterparty_id) {
      result.push_back(limit);
    }
  }
  return result;
}

Status CounterpartyLimit::CheckLimit(const ExposureLimit& limit,
                                     const ExposureReading& reading,
                                     std::uint32_t record_id,
                                     LimitBreach* out) const {
  if (out == nullptr) {
    return Status::kBoundsError;
  }

  std::int64_t current = 0;
  switch (limit.kind) {
    case LimitKind::kNotional:
      current = reading.notional_cents;
      break;
    case LimitKind::kQuantity:
      current = reading.qty_milli;
      break;
    case LimitKind::kDailyTurnover:
      current = reading.daily_turnover_cents;
      break;
    case LimitKind::kNetExposure:
      current = reading.net_exposure_cents;
      break;
    case LimitKind::kOpenOrders:
      current = reading.open_order_count;
      break;
    default:
      return Status::kOk;
  }

  if (limit.scope == LimitScope::kPerSymbol &&
      limit.symbol_id != 0 && limit.symbol_id != reading.symbol_id) {
    return Status::kOk;
  }

  if (limit.scope == LimitScope::kPerAccount &&
      limit.account_id != 0 && limit.account_id != reading.account_id) {
    return Status::kOk;
  }

  const std::int64_t prior =
      LookupCurrentExposure(reading.counterparty_id, limit.kind);
  current += prior;

  if (current > limit.max_value) {
    out->kind = limit.kind;
    out->limit_id = limit.limit_id;
    out->counterparty_id = reading.counterparty_id;
    out->record_id = record_id;
    out->current_value = current;
    out->limit_value = limit.max_value;
    out->is_warning = false;
    out->reason = "limit breach: current " + std::to_string(current) +
                  " exceeds max " + std::to_string(limit.max_value);
    return limit.hard_block ? Status::kComplianceReject : Status::kOk;
  }

  if (current > limit.warning_threshold) {
    out->kind = limit.kind;
    out->limit_id = limit.limit_id;
    out->counterparty_id = reading.counterparty_id;
    out->record_id = record_id;
    out->current_value = current;
    out->limit_value = limit.warning_threshold;
    out->is_warning = true;
    out->reason = "limit warning: approaching threshold";
  }

  return Status::kOk;
}

CounterpartyLimitResult CounterpartyLimit::EvaluateReading(
    const ExposureReading& reading) {
  CounterpartyLimitResult result{};
  result.status = Status::kOk;

  if (!IsCounterpartyApproved(reading.counterparty_id)) {
    LimitBreach breach{};
    breach.kind = LimitKind::kNotional;
    breach.counterparty_id = reading.counterparty_id;
    breach.is_warning = false;
    breach.reason = "counterparty not approved";
    result.breaches.push_back(breach);
    ++result.breach_count;
    result.status = Status::kComplianceReject;
    return result;
  }

  const std::vector<ExposureLimit> applicable =
      LimitsForCounterparty(reading.counterparty_id);

  for (const ExposureLimit& limit : applicable) {
    LimitBreach breach{};
    Status st = CheckLimit(limit, reading, 0, &breach);
    if (st == Status::kComplianceReject) {
      result.breaches.push_back(breach);
      ++result.breach_count;
      if (config_.enforce_hard_blocks) {
        result.status = Status::kComplianceReject;
      }
    } else if (breach.is_warning) {
      result.breaches.push_back(breach);
      ++result.warning_count;
    }
  }

  ExposureReading& running = running_exposure_[reading.counterparty_id];
  running.counterparty_id = reading.counterparty_id;
  running.notional_cents += reading.notional_cents;
  running.qty_milli += reading.qty_milli;
  running.daily_turnover_cents += reading.daily_turnover_cents;
  running.net_exposure_cents += reading.net_exposure_cents;
  running.open_order_count += reading.open_order_count;

  result.readings.push_back(reading);
  return result;
}

CounterpartyLimitResult CounterpartyLimit::Evaluate(
    const BatchWireFrame& frame) {
  CounterpartyLimitResult result{};
  result.status = Status::kOk;

  std::vector<ExposureReading> batch_readings;
  batch_readings.reserve(frame.records.size());

  for (const WireBatchRecord& rec : frame.records) {
    batch_readings.push_back(BuildReading(rec));
  }

  std::unordered_map<std::uint32_t, ExposureReading> aggregated;
  for (const ExposureReading& r : batch_readings) {
    ExposureReading& agg = aggregated[r.counterparty_id];
    agg.counterparty_id = r.counterparty_id;
    agg.notional_cents += r.notional_cents;
    agg.qty_milli += r.qty_milli;
    agg.daily_turnover_cents += r.daily_turnover_cents;
    agg.net_exposure_cents += r.net_exposure_cents;
    agg.open_order_count += r.open_order_count;
  }

  for (const WireBatchRecord& rec : frame.records) {
    ExposureReading reading = BuildReading(rec);
    CounterpartyLimitResult partial = EvaluateReading(reading);
    for (const LimitBreach& b : partial.breaches) {
      result.breaches.push_back(b);
      if (b.is_warning) {
        ++result.warning_count;
      } else {
        ++result.breach_count;
      }
    }
    if (partial.status == Status::kComplianceReject &&
        config_.enforce_hard_blocks) {
      result.status = Status::kComplianceReject;
    }
    result.readings.push_back(reading);
  }

  if (result.breach_count > 0 && config_.enforce_hard_blocks) {
    result.status = Status::kComplianceReject;
  }

  return result;
}

CounterpartyLimitResult CounterpartyLimit::EvaluateParentAggregate(
    const BatchWireFrame& frame) {
  CounterpartyLimitResult result{};
  result.status = Status::kOk;

  if (!config_.aggregate_parent) {
    return Evaluate(frame);
  }

  std::unordered_map<std::uint32_t, std::vector<ExposureReading>> by_parent;

  for (const WireBatchRecord& rec : frame.records) {
    ExposureReading reading = BuildReading(rec);
    std::uint32_t parent_id = reading.counterparty_id;
    const auto cp_it = counterparties_.find(reading.counterparty_id);
    if (cp_it != counterparties_.end()) {
      parent_id = cp_it->second.parent_id ? cp_it->second.parent_id
                                          : cp_it->second.counterparty_id;
    }
    by_parent[parent_id].push_back(reading);
  }

  for (auto& pair : by_parent) {
    ExposureReading agg = AggregateReadings(pair.second, pair.first);
    CounterpartyLimitResult partial = EvaluateReading(agg);
    for (const LimitBreach& b : partial.breaches) {
      result.breaches.push_back(b);
      if (b.is_warning) {
        ++result.warning_count;
      } else {
        ++result.breach_count;
      }
    }
    result.readings.push_back(agg);
  }

  if (result.breach_count > 0 && config_.enforce_hard_blocks) {
    result.status = Status::kComplianceReject;
  }
  return result;
}

void CounterpartyLimit::ResetRunningExposure() {
  running_exposure_.clear();
}

std::int64_t CounterpartyLimit::TotalDeskExposure(LimitKind kind) const {
  std::int64_t total = 0;
  for (const auto& pair : running_exposure_) {
    switch (kind) {
      case LimitKind::kNotional:
        total += pair.second.notional_cents;
        break;
      case LimitKind::kQuantity:
        total += pair.second.qty_milli;
        break;
      case LimitKind::kDailyTurnover:
        total += pair.second.daily_turnover_cents;
        break;
      default:
        break;
    }
  }
  return total;
}

CounterpartyProfile CounterpartyLimit::LookupProfile(
    std::uint32_t counterparty_id) const {
  const auto it = counterparties_.find(counterparty_id);
  if (it == counterparties_.end()) {
    return CounterpartyProfile{};
  }
  return it->second;
}

}  // namespace desk
}  // namespace tkr
